flow/src/main/java/com/sequenceiq/flow/core/Flow2Handler.java

```diff
@@ -439,6 +443,25 @@ public FlowIdentifier retryLastFailedFlow(Long resourceId, java.util.function.Co
                 .orElseThrow(() -> new BadRequestException("Retry cannot be performed, because the last action was successful."));
     }
 
+    /**
+     * Retry the failed flow completely
+     * @param resourceId Datalake ID
+     * @return Identifier of flow or a flow chain
+     */
+    public FlowIdentifier retryLastFailedFlowFromStart(Long resourceId) {
+        List<FlowLog> flowLogs = flowLogService.findAllForLastFlowIdByResourceIdOrderByCreatedDesc(resourceId);
+        canRetryBePerformed(flowLogs);
+        FlowLog firstSuccessfulStateLog = FlowLogUtil.getFirstStateLog(flowLogs);
+        LOGGER.info("Trying to restart flow {}", firstSuccessfulStateLog.getFlowType().getName());
+        restartFlow(firstSuccessfulStateLog);
+        LOGGER.info("Restarted flow : {}", firstSuccessfulStateLog.getFlowType().getName());
+        if (firstSuccessfulStateLog.getFlowChainId() != null) {
+            return new FlowIdentifier(FlowType.FLOW_CHAIN, firstSuccessfulStateLog.getFlowChainId());
```

Comment:
Fortunately it works. Flow chains can be arbitrarily nested but the `com.sequenceiq.flow.api.FlowPublicEndpoint#hasFlowRunningByChainId` can determine if the root flow chain is completed or not based on child flow chain ids.

Reply:
I do not understand your comment.

Reply:
Are you suggesting any change here?